zerver/migrations/0417_backfill_remote_realmauditlog_extradata_to_json_field.py

```diff
@@ -0,0 +1,40 @@
+# Generated by Django 4.0.7 on 2022-09-30 20:30
+
+import orjson
+import yaml
+from django.db import migrations
+from django.db.backends.postgresql.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+
+def backfill_extra_data(apps: StateApps, schema_editor: BaseDatabaseSchemaEditor) -> None:
+    RealmAuditLog = apps.get_model("zerver", "RealmAuditLog")
+
+    for audit_log_entry in RealmAuditLog.objects.filter(extra_data__isnull=False):
+        if audit_log_entry.event_type in (audit_log_entry.USER_FULL_NAME_CHANGED,):
+            # extra_data is a plain str, containing the old full name
+            extra_data_json = {"old_name": audit_log_entry.extra_data}
+        else:
+            try:
+                extra_data_json = orjson.loads(audit_log_entry.extra_data)
+            except orjson.JSONDecodeError:
+                # Extra data might contain improperly deserialized data generated from code like str({"foo": "bar"})
+                # that cannot be parsed by orjson. An examples of this is do_change_remote_server_plan_type.
+                # The yaml SafeLoader is capable of handling this situation and properly generates a dict
+                # for backfilling.
+                extra_data_json = yaml.safe_load(audit_log_entry.extra_data)
+        audit_log_entry.extra_data_json = extra_data_json
+        audit_log_entry.save(update_fields=["extra_data_json"])
```

Comment:
Round-tripping every row to Python is almost certainly going to be too slow.

Perhaps we can filter for rows that don’t start with `{'` and do that work on the PostgreSQL server: for example,

```python
from django.db.models import JSONField
from django.db.models.functions import Cast

RealmAuditLog.objects.exclude(extra_data__startswith="{'").update(
    extra_data_json=Cast("extra_data", output_field=JSONField())
)
```

is translated to the PostgreSQL query

```sql
UPDATE "zerver_realmauditlog"
SET "extra_data_json" = ("zerver_realmauditlog"."extra_data")::jsonb
WHERE NOT ("zerver_realmauditlog"."extra_data"::text LIKE '{''%' AND "zerver_realmauditlog"."extra_data" IS NOT NULL)
```

Then we can do another query with `.filter(extra_data__startswith="{'")`, process those rows in Python, and update them using [`bulk_update`](https://docs.djangoproject.com/en/4.1/ref/models/querysets/#bulk-update) to avoid the round-trip-per-row.

To avoid a large transaction that might block the server for significant time (although that may not be a concern for `RealmAuditLog` specifically), we can batch these queries by ID (`.filter(id__range=(low, high))`).

Reply:
This gets added in https://github.com/zulip/zulip/pull/23116/commits/8de631a629cc4e886352bbf78fa255ee00880dd9.